File: Java/src/main/java/org/codecop/lcdnumbers/DigitPrinter.java
package org.codecop.lcdnumbers;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Appends lines next to each other and adds line breaks.
 */
public class DigitPrinter {

    private static final String NEWLINE = "\n";

    private final Zipper zipper;

    public DigitPrinter(Zipper zipper) {
        Objects.requireNonNull(zipper);

        this.zipper = zipper;
    }

    public String render(List<ScaledDigit> digits) {
        Objects.requireNonNull(digits);

        List<List<ScaledLine>> linesOfAllDigits = linesOfAllDigits(digits);
        List<String> linesSideBySide = zip(linesOfAllDigits);
        return join(linesSideBySide);
    }

    private List<List<ScaledLine>> linesOfAllDigits(List<ScaledDigit> digits) {
        return digits.stream(). //
                map(ScaledDigit::lines). //
                collect(Collectors.toList());
    }

    private List<String> zip(List<List<ScaledLine>> linesOfAllDigits) {
        return zipper.zip(linesOfAllDigits, this::concat);
    }

    private String concat(List<ScaledLine> lines) {
        return lines.stream(). //
                map(ScaledLine::toString). //
                collect(Collectors.joining());
    }

    private String join(List<String> lines) {
        return lines.stream(). //
                collect(Collectors.joining(NEWLINE, "", NEWLINE));
    }

}

File: JavaScript/src/digit_printer.js
exports = typeof window !== 'undefined' && window !== null ? window : global;

/**
 * Appends lines next to each other and adds line breaks.
 */
exports.DigitPrinter = function(zip) {

    var NEWLINE = "\n";

    // TODO Objects.requireNonNull(zip);

    function linesOfAllDigits(digits) {
        return digits.map(function(digit) { return digit.lines(); });
    }

    function doZip(linesOfAllDigits) {
        return zip(linesOfAllDigits, concat);
    }

    function concat(lines) {
        return lines.map(function(line) { return line.toString(); }).
                reduce(function (a, b) { return a + b; });
    }

    function join(lines) {
        return lines.reduce(function (a, b) { return a + NEWLINE + b ; }) + NEWLINE;
    }

    this.render = function(digits) {
        // TODO Objects.requireNonNull(digits);

        var _linesOfAllDigits = linesOfAllDigits(digits);
        var linesSideBySide = doZip(_linesOfAllDigits);
        return join(linesSideBySide);
    };

};
